ledger-api/grpc-definitions/com/daml/ledger/api/v1/version_service.proto

```diff
@@ -26,6 +28,106 @@ message GetLedgerApiVersionRequest {
 
 message GetLedgerApiVersionResponse {
 
-  // The version of the ledger API
+  // The version of the ledger API.
   string version = 1;
+
+  // The features supported by this Ledger API endpoint.
+  //
+  // Daml applications CAN use the feature descriptor on top of 
+  // version constraints on the Ledger API version to determine
+  // whether a given Ledger API endpoint supports the features
+  // required to run the application.
+  //
+  // See the feature descriptions themselves for the relation between
+  // Ledger API versions and feature presence.
+  FeatureDescriptor features = 2;
+}
+
+message FeatureDescriptor {
+  // Features that are determined by the ledger backing the participant node
+  // that serves the Ledger API endpoint.
+  //
+  // All participant nodes connected to the same ledger are expected to report the
+  // same values for the ledger features.
+  LedgerFeatures ledger = 1;
+
+  // Features that are are specific to the participant node that serves
+  // the Ledger API endpoint. 
+  ParticipantFeatures participant = 2;
+
+  // Features under development or features that are used
+  // for ledger implementation testing purposes only.
+  //
+  // Daml applications SHOULD not depend on these in production.
+  ExperimentalFeatures experimental = 3;
+}
+
+// See the feature message definitions for descriptions.
+message LedgerFeatures {
+  FeatureOrderedContractIds ordered_contract_ids = 1;
+  FeatureMonotonicRecordTime monotonic_record_time = 2;  
```

Comment:
Perhaps we can explicitly mark each of these as `optional` (and below)?

Reply:
Good idea. Will do.

Reply:
Actually, it does just add noise doesn't it? `optional` is the default in proto3, and we generally do not mark fields as optional do we?

Reply:
We normally mark optional fields because our convention is that all fields are required.